### PR TITLE
txn: add watermark

### PR DESCRIPTION
**Pull request by GanZiheng**
Based on
- https://github.com/tikv/agatedb/pull/57
- https://github.com/dgraph-io/badger/blob/45bca18f24ef5cc04701a1e17448ddfce9372da0/y/watermark.go

Signed-off-by: GanZiheng [ganziheng98@gmail.com](mailto:ganziheng98@gmail.com)

### CODE REVIEW

**Comment by GanZiheng** [comment]
Will rebase onto master once https://github.com/tikv/agatedb/pull/129 merged.

**Comment by GanZiheng** [comment]
A little bit weird, we will get thread sanitizer error if we use yatp. For example, https://github.com/tikv/agatedb/runs/6705381794?check_suite_focus=true.

**Comment by GanZiheng** [comment]
> 

So I use standard library to spawn threads.

**Comment by skyzh** [comment]
> https://github.com/tikv/agatedb/runs/6705381794?check_suite_focus=true

This is a known issue of crossbeam-epoch. If we want to use crossbeam channel in AgateDB, we will need to disable thread sanitizer check (or allow it to ignore crossbeam-related structures).

Few projects will run tests with sanitizers. I think we can periodically check address / leak / thread sanitizer when things go wrong, and simply remove sanitizer checks from CI.

**Comment by GanZiheng** [comment]
> > https://github.com/tikv/agatedb/runs/6705381794?check_suite_focus=true
> 
> This is a known issue of crossbeam-epoch. If we want to use crossbeam channel in AgateDB, we will need to disable thread sanitizer check (or allow it to ignore crossbeam-related structures).
> 
> Few projects will run tests with sanitizers. I think we can periodically check address / leak / thread sanitizer when things go wrong, and simply remove sanitizer checks from CI.

But when I use standard library to spawn thread, the issue disappears. Sort of strange...

**Comment by skyzh** [comment]
> But when I use standard library to spawn thread, the issue disappears. Sort of strange...

Maybe by accident or some magic 🤪

**Comment by zhangjinpeng1987** [comment]
@GanZiheng Please fix conflicts.

File: src/watermark.rs [hunk 1]
```diff
@@ -0,0 +1,340 @@
...
+// status of the index: begin or done. It also contains waiters, who could be
+// waiting for the watermark to reach >= a certain index.
+#[derive(Debug)]
+struct Mark {
+    index: Index,
+    waiter: Option<Sender<()>>,
+    done: bool,
+}
+
+struct Core {
```
Comment (Connor1996) [inline src/watermark.rs:32]: rename it to `WaterMarkInner`

File: src/watermark.rs [hunk 2]
```diff
@@ -164,15 +164,15 @@ impl Core {
 /// number of threads needed.
 #[derive(Clone)]
 pub struct WaterMark {
-    core: Arc<Core>,
+    core: Arc<WaterMarkInner>,
```
[the remark this answers is not in the document]
Comment (GanZiheng) [reply src/watermark.rs:167]: Oops, forgot it, will fix it soon.

File: src/closer.rs [hunk 3]
```diff
@@ -29,30 +29,31 @@ impl Closer {
...
 
         let (tx, rx) = unbounded::<()>();
 
+        let mut handles = vec![];
+
         for _i in 0..10 {
             let closer = closer.clone();
             let tx = tx.clone();
-            pool.spawn(move |_: &mut Handle<'_>| {
+            handles.push(std::thread::spawn(move || {
```
Comment (zhangjinpeng1987) [inline src/closer.rs:45]: Does this change is want to workaround the thread sanitizer? If so, master has fix it.

Comment (GanZiheng) [reply src/closer.rs:45]: I will change it back.

**Review by zhangjinpeng1987 — APPROVED** [review]
LGTM

**Pull request merged**